### DISCUSSION TITLE
how to add things dynamically (and correctly)?

### DISCUSSION THREAD

**Discussion by marcin-koziol** [zauberzeug/nicegui#2013] [Q&A] [answered]
### Question

So, here's my issue:

I want to create a dynamic formset with "add" button (yes, a Django's nomenclature again). Tables or aggrid are not a solution, because I need more complex things to add (input elements, buttons for invoking popups etc), so currently, I'm adding rows to a list and then, using @refreshable decorator I'm almost done, but, depending on single form complexity, after adding 20-30 rows, refreshing starts to take noticeably longer. Is there any other way to "add things"?

**Comment by falkoschindler** [comment, marked as answer]
Yes, you can simply create new UI elements. Make sure to enter the right context using `with` expressions:
```py
container = ui.column()

def add():
    with container:
        with ui.row() as row:
            ui.label(f'Text {len(container.default_slot.children)}')
            ui.button('Remove', on_click=lambda: container.remove(row))

ui.button('Add', on_click=add)
```

It is a bit more work than using `ui.refreshable`, but this way you don't re-create and transmit the whole UI every time. If the number of elements grows and performance is an issue, adding and removing individual elements is definitely the way to go.

**Comment by marcin-koziol** [comment]
And what about reordering rows?
ahh, nevermind, I can just swap values

**Reply by falkoschindler** [reply]
And you can [move elements](https://nicegui.io/documentation/element#move_elements).
